tequilapi/endpoints/stop.go

```diff
@@ -4,28 +4,25 @@ import (
 	log "github.com/cihub/seelog"
 	"github.com/julienschmidt/httprouter"
 	"net/http"
-	"time"
 )
 
 // ApplicationStopper stops application and performs required cleanup tasks
 type ApplicationStopper func()
 
 // AddRouteForStop adds stop route to given router
-func AddRouteForStop(router *httprouter.Router, stop ApplicationStopper, delay time.Duration) {
-	router.POST("/stop", newStopHandler(stop, delay))
+func AddRouteForStop(router *httprouter.Router, stop ApplicationStopper) {
+	router.POST("/stop", newStopHandler(stop))
 }
 
-func newStopHandler(stop ApplicationStopper, delay time.Duration) httprouter.Handle {
-	return func(response http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
-		log.Info("Client is stopping")
-		delayExecution(stop, delay)
+func newStopHandler(stop ApplicationStopper) httprouter.Handle {
+	return func(response http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+		log.Info("Application stop requested")
+
+		go doWhenNotified(req.Context().Done(), stop)
 		response.WriteHeader(http.StatusAccepted)
 	}
 }
-
-func delayExecution(function func(), duration time.Duration) {
-	go func() {
-		time.Sleep(duration)
-		function()
-	}()
+func doWhenNotified(notify <-chan struct{}, stopApplication ApplicationStopper) {
```

Comment:
Function name is generic `doWhenNotified`, but param is concrete - `stopApplication`. Can this function perform other things when notified?

Reply:
Well it's to private and two-linear to do other things. Maybe callStopWhenNotified?

Reply:
>to private and two-linear
you mean "too private and too linear"? :D
For me this looks like a generic thing - this was the original intention. Of course it's private, so you can't re-use it now, but in case you had to, it's possible to extract it - re-usable components are healthy even when not re-using :)

Reply:
I mean "too much internal" and two-(2 lines)-linear.  Yea I will stick with renaming. There is an old saying - don't DRY too much, you will suffer from dehidratation :)